gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/GetProjectInvocations.java

```diff
@@ -0,0 +1,134 @@
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.google.errorprone.suppliers.Supplier;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.util.TreeScanner;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.Type.ClassType;
+import java.util.List;
+
+@AutoService(BugChecker.class)
+@BugPattern(severity = SeverityLevel.ERROR, summary = GetProjectInvocations.SUMMARY)
+public final class GetProjectInvocations extends GradleGuideBugChecker implements BugChecker.MethodTreeMatcher {
+    private static final Supplier<ClassSymbol> ACTION_SYM =
+            VisitorState.memoize(s -> (ClassSymbol) s.getSymbolFromString("org.gradle.api.Action"));
+    private static final Supplier<ClassSymbol> TASK_SYM =
+            VisitorState.memoize(s -> (ClassSymbol) s.getSymbolFromString("org.gradle.api.Task"));
+
+    private static final Matcher<ExpressionTree> TASK_GET_PROJECT_METHOD = MethodMatchers.instanceMethod()
+            .onDescendantOf("org.gradle.api.Task")
+            .named("getProject");
+
+    public static final String VIOLATION_MESSAGE = "Don't call `getProject()` in task actions";
+
+    public static final String SUMMARY =
+            """
+        Don't call `getProject()` in task actions. Large, mutable Gradle model types like `Gradle`, `Settings`, or
+        `Project` should not be passed into tasks as inputs. Instead, your tasks should take in the "smallest" type
+        required for the task's functionality. For example, instead of taking in `Project` to later do
+        `project.version`, you should declare the project version as a `Property<String>`.
+        Doing so improves performance in two ways:
+        1. It makes your tasks compatible with the configuration cache
+        2. It prevents tasks from being rendered out-of-date by a mutation unrelated to the task, e.g. to `project.name`
+        """;
+
+    @Override
+    public Description matchMethod(MethodTree tree, VisitorState state) {
+        if (isTaskAction(tree, state) || overridesExecute(tree, state)) {
+            reportAllViolations(tree.getBody(), state, TASK_GET_PROJECT_METHOD, VIOLATION_MESSAGE);
+        }
+
+        return Description.NO_MATCH;
+    }
+
+    private static boolean isTaskAction(MethodTree tree, VisitorState state) {
+        return tree.getModifiers().getAnnotations().stream().anyMatch(annotation -> {
+            String source = state.getSourceForNode(annotation.getAnnotationType());
+            return source.equals("TaskAction");
+        });
```

Comment:
We should avoid looking at source where ever possible. It's sometimes not available for one.

Instead I think we can just use `Matchers.hasAnnotation("org.gradle.api.tasks.TaskAction")`

Reply:
Gotcha, I'll make the change.

> It's sometimes not available for one

When so?

> it's also less precise than looking at the AST

Seems like Matchers.hasAnnotation matches with Symbols. Is that generally what we want to do, for precision?